corehq/apps/hqadmin/service_checks.py

```diff
@@ -230,6 +233,29 @@ def check_formplayer():
         return ServiceStatus(res.ok, msg)
 
 
+def check_pillows():
+    pillows = [
+        pillow for pillow in get_all_pillow_instances()
+        if pillow.pillow_id in getattr(settings, 'ACTIVE_PILLOW_NAMES', [pillow.pillow_id])
+        and isinstance(pillow.get_change_feed(), KafkaChangeFeed)
+    ]
+
+    failed_pillows = []
+    for pillow in pillows:
+        oldest_checkpoint = min(c.last_modified for c in pillow.checkpoint._get_checkpoints())
+        time_since_oldest = datetime.datetime.utcnow() - oldest_checkpoint
+
+        # allow 60 second buffer since save only occurs after max is exceeded
+        # and to account for delay between read and comparison
+        if time_since_oldest.seconds > (MAX_CHECKPOINT_DELAY + 60):
```

Comment:
IIRC I believe that the checkpoint isn't forced if there are no updates. Meaning that if MAX_CHECKPOINT_DELAY is 5 minutes, and there haven't been any changes in 10 minutes, this failure would report as failed. This might be fine on ICDS, but it might make other envs much noisier, especially monoliths.

I think there are probably a couple of ways to guard against it.

A) Sending a 'heartbeat' change through kafka that is ignored by pillows, but would increase the counter
B) Setting a timeout on the change feeds so that if there isn't an activity for X seconds, the loop exits and we can save the checkpoint
C) Changing this check to be a combination of checking the timestamp and number of changes away from the offset.

Reply:
Oof. I should have read your comment and full description before posting this, but hopefully the options help out

Reply:
Haha, yeah. I decided on C short after writing that comment, but got pulled into other things before I could finish. Am picking up from there tomorrow, I think its pretty close